### PR TITLE
schemadiff: load, normalize, validate and compare schemas

### PR DESCRIPTION
**Pull request by shlomi-noach**
## Description

Up till now [schemadiff](https://github.com/vitessio/vitess/pull/9719) allowed us to compare a table with another table, or a view with another view. This is [now in use](https://github.com/vitessio/vitess/pull/9772) in declarative schema changes.

This PR introduces schema comparisons.

A `Schema` is a collection of database objects -- right now this means tables and views. You may load a schema by:

- a list of well formed entities, or
- a list of well formed `Statement`s, or
- a list of `CREATE ...` queries, or
- a SQL blob that contains multiple `CREATE` queries

A schema is validated upon loading:

- can't have two objects with same name
- all view dependencies are resolved

As for the latter: `schemadiff` parses each view and analyze its dependencies. It validates that all dependencies are met. e.g. you can't create a view that reads from a nonexistent table. Also, you can't create two views that have circular dependency.

Lastly, the schema is normalized: we reorder the objects by order of valid creation: first, the tables; then, views that only depend on tables; then, 2nd level views, etc.

### Diff

`schemadiff` can now do, well, a schema-diff. Compare two `Schema` objects. This returns a (possibly empty) list of diffs, which create/alter/drop tables and views so as to mutate one schema to another. This is all done declaratively.

See the added tests for more insights.

## Checklist
- [ ] Should this PR be backported?
- [x] Tests were added or are not required
- [ ] Documentation was added or is not required

### CODE REVIEW

File: go/vt/schemadiff/diff.go [hunk 1]
```diff
@@ -116,15 +113,46 @@ func DiffViews(create1 *sqlparser.CreateView, create2 *sqlparser.CreateView, hin
...
+	}
+	return schema1.Diff(schema2, hints)
+}
+
+// DiffSchemasSQL compares two schemas and returns the list of diffs that turn
+// 1st schema into 2nd. Any of the schemas may be nil.
+func DiffSchemas(schema1 *Schema, schema2 *Schema, hints *DiffHints) ([]EntityDiff, error) {
+	var err error
+	if schema1 == nil {
+		schema1, err = NewSchemaFromStatements(nil)
```
Comment (vmg) [inline go/vt/schemadiff/diff.go:146]: Is there a reason why this isn't `newEmptySchema` directly?

**Review by vmg — APPROVED** [review]
This looks great. I only have a small nit. :sparkles:

File: go/vt/schemadiff/diff.go (hunk 1 above)
Comment (shlomi-noach) [reply go/vt/schemadiff/diff.go]: Good catch! Now simplified.

**Comment by shlomi-noach** [comment]
Tracking issue: https://github.com/vitessio/vitess/issues/10203

**Pull request merged**